Payment Request Button: Guide users to authentication when it is required

ricardo opened the pull request:
Fixes #1502

#### Changes proposed in this Pull Request

- Display the button instead of hiding it when authentication is required.
- Add a dialog which tells users to login before checking out.
- Redirect users back to the checkout flow after authentication.

#### Testing instructions

First, run `npm start` to build the assets from this branch.

1. Go to **WooCommerce > Settings > Accounts & Privacy** and disable the following settings:
  - Allow customers to place orders without an account.
  - Allow customers to create an account during checkout.
2. Make sure you have **Payment request buttons** enabled under **Payments > Settings**.
3. Make sure you're not logged in.
3. Go to a product page in Chrome and click the payment request button.
4. Notice there's a dialog and the "Confirm" button redirects you to my-account.
5. Sign in or create an account.
6. After authentication, you should be redirected back to the same page.

![gif](https://user-images.githubusercontent.com/5509901/122165183-3a8ca980-ce4e-11eb-9d18-b328edc74935.gif)


-------------------

- [x] Added changelog entry (or does not apply)
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

asumaran commented:
@ricardo I'm unable to test this. When I click the subscription button it goes to the Cart.

I've tested it with a non-logged in user and a subscription created with WooCommerce Subscriptions.  If it helps, `wcpayPaymentRequestParams` is not defined. Let me know if you need more details on how I tested it.

asumaran commented on client/payment-request/redirect-dialog.js:
```diff
@@ -0,0 +1,38 @@
+	// For reference, check `tb_show` and `tb_position` in WP core thickbox.js.
+	// Note: We shouldn't change the CSS for the default Thickbox elements,
+	// otherwise it may break the styling in other pages.
+	jQuery( 'body' ).append( '<div class="TB_wrapper"></div>' );
```
I'm not familiar but I guess WCPay already loads jQuery so we use jQuery to interact with the DOM by default?

asumaran commented on includes/class-wc-payments-payment-request-button-handler.php:
```diff
@@ -175,6 +179,24 @@ public function set_session() {
+			$url = rawurldecode( esc_url_raw( wp_unslash( $_GET['wcpay_payment_request_redirect_url'] ) ) );
+			// Sets a redirect URL cookie for 10 minutes,
+			// which we will redirect to after authentication.
+			wc_setcookie( 'wcpay_payment_request_redirect_url', $url, time() + MINUTE_IN_SECONDS * 10 );
```
What happens if the 10 minutes passes and the user logs in after that?

ricardo replied:
If 10 minutes passes they're not redirected to the same page after login. This 10 minute cookie is an arbitrary timeout to login and complete the PRB transaction.

asumaran commented on includes/class-wc-payments-payment-request-button-handler.php:
```diff
@@ -560,6 +577,29 @@ public function get_product() {
+	 * @return string Redirect URL.
+	 */
+	public function get_login_redirect_url( $redirect ) {
+		$url = esc_url_raw( wp_unslash( $_COOKIE['wcpay_payment_request_redirect_url'] ?? '' ) );
```
is this the way we do it? I was thinking maybe the standard way was to pass the URL via query string. like we do when an unauthenticated user tries to go to a page in the admin area: 
`http://domain.tld/wp-login.php?redirect_to=http%3A%2F%2Fwcpay.jurassic.tube%2Fwp-admin%2Fadmin.php%3Fpage%3Dwc-admin&reauth=1`

I guess since we are using `woocommerce_login_redirect` filter this is how we do it.

ricardo replied:
To redirect after login/registration within WooCommerce from the "my account" page, the only way I found was through the filters `woocommerce_login_redirect` and `woocommerce_registration_redirect`.

asumaran commented on client/payment-request/style.scss:
```diff
@@ -1,4 +1,23 @@
+	pointer-events: none;
+	// Allow pointer events in inner window only, so the overlay
+	// is still clickable and allows for closing the dialog.
+	div {
```
maybe safer than this is `> *` but this one works too

ricardo replied:
I actually meant to be more specific about the selector here. Fixed in 877e92b7cde17bb172fcb8737ba37c0a651eac6d.

asumaran reviewed (changes requested):
@ricardo I left a few comments but as I mentioned I wasn't able to test this PR. I'll do it a second pass once the issue I'm having is fixed.

ricardo commented on client/payment-request/blocks/use-initialization.js:
```diff
@@ -69,23 +72,38 @@ export const useInitialization = ( {
+				total: billing?.cartTotal?.value,
+				displayItems: normalizeLineItems( billing?.cartTotalItems ),
+			} );
+			onClick();
```
`onClick` is a prop from `PaymentRequestExpressComponent`, which is passed to the `useInitialization` hook. It's defined by the express payment method.

ricardo commented on client/payment-request/index.js:
```diff
@@ -167,7 +171,7 @@ jQuery( ( $ ) => {
 			} );
 
-			api.paymentRequestAddToCart( data );
+			return api.paymentRequestAddToCart( data );
```
I noticed the return was missing and I added it with plans for 1563, but it's not really necessary for this PR.

ricardo commented:
@asumaran I have changed the behavior for subscription products and now it should open the redirect dialog if account creation is not possible during checkout (meaning `Allow customers to create an account during checkout` is disabled or automatically generating and account username/password are disabled).

Also implemented the design feedback from p1620857968351700-slack-CGGCLBN58 with some tweaks on the dialog messaging and removed the automatic initialization of the payment request after redirect.

This implementation should also work for the block cart. Let me know if there's something left for me to do regarding your previous comments, or if you're unable to test the PR.

asumaran commented on includes/class-wc-payments-payment-request-button-handler.php:
```diff
@@ -496,6 +512,33 @@ public function has_allowed_items_in_cart() {
+		} elseif ( $this->is_checkout() || $this->is_cart() ) {
+			foreach ( WC()->cart->get_cart() as $cart_item_key => $cart_item ) {
+				$_product = apply_filters( 'woocommerce_cart_item_product', $cart_item['data'], $cart_item, $cart_item_key );
+				if ( WC_Subscriptions_Product::is_subscription( $_product ) ) {
+					return true;
+				}
```
So, If I'm reading this correctly we will use this method to tell the user they need to log in if at least one subscription is present on the cart right? In that case are we aware the user will not be able to purchase other products if a subscription is present in the cart? My point is, we are not specifically telling the user they can not purchase the other products due to a subscription being present in the cart.

The message in the cart doesn't explicitly tells the user the subscription is the one that requires the user to sign in to be able to purchase the products.

ricardo replied:
> if at least one subscription is present on the cart right?

Exactly, and if it's not possible to create an account upon checkout.

> The message in the cart doesn't explicitly tells the user the subscription is the one that requires the user to sign in to be able to purchase all products.

I believe we should view this as: "Can I buy the items on the cart without having an account?" If not, I should be redirected to the login page. I guess if we say something about a specific product being the blocker, we could implicitly tell people to remove the subscription in order to proceed with the order. 🤔

We don't want users to give up on a subscription from their cart just because they need to create an account first, so in my opinion the message is fine.

I didn't personally craft the messaging, but we can ask for a design opinion considering this scenario. Do you think we should?

asumaran replied:
> "Can I buy the items on the cart without having an account?" If not, I should be redirected to the login page.

I see what you mean now. If the copy was made considering the case I'm presenting then looks good to me.

ricardo replied:
Just confirmed with David and the message covers both cases. Marking as resolved ✅

asumaran commented on includes/class-wc-payments-payment-request-button-handler.php:
```diff
@@ -562,6 +605,23 @@ public function get_product() {
+	 * @return string Redirect URL.
+	 */
+	public function get_login_redirect_url( $redirect ) {
+		$url = esc_url_raw( wp_unslash( $_COOKIE['wcpay_payment_request_redirect_url'] ?? '' ) );
```
This looked a little bit risky. A bad intentioned user could set this cookie via JS or in the backend and redirect to another site. I've tried changing the value to `https://google.com` in the cookie but it seems there we already have protections to avoid unwanted redirects.

ricardo replied:
> A bad intentioned user could set this cookie via JS or in the backend

In the backend do you mean via a malicious plugin? In that case we would be vulnerable everywhere and I don't think we can avoid it, but it's a different story when it's caused by the website administrators. What we definitely can and always should avoid is attacks from third parties.

Setting the cookie on the front-end or via JS would affect only the user who has set the cookie, not others, unless I create a browser extension to set this cookie to others.

Do you have suggestions on how to implement this differently? If this could get changed to some other URL, what bad things could happen do you foresee?

asumaran replied:
> Do you have suggestions on how to implement this differently? If this could get changed to some other URL, what bad things could happen do you foresee?

it's OK to use this method since I wasn't able to redirect to another domain even when modifying the cookie manually.

asumaran reviewed (changes requested):
Overall looks good. I left a few comments for you @ricardo

ricardo commented on includes/class-wc-payments-payment-request-button-handler.php:
```diff
@@ -658,6 +720,44 @@ public function display_payment_request_button_separator_html() {
+	/**
+	 * Display payment request redirect dialog.
+	 */
+	public function display_payment_request_redirect_dialog() {
```
Well observed. This was actually considered and discussed before (p1615518244176400/1615427515.099500-slack-CGGCLBN58).

The `Anyone can register` option doesn't really apply in this case because it's implicit that to use the PRB you'd need to have that option checked. Any issues with having it disabled would be noticeable when testing. So we don't need to change the messaging.

ricardo commented on includes/class-wc-payments-payment-request-button-handler.php:
```diff
@@ -658,6 +720,44 @@ public function display_payment_request_button_separator_html() {
+		);
+		?>
+		<div id="payment-request-redirect-dialog" style="display:none;">
+			<div style="padding: 1.5rem 0">
```
I followed the same approach from other elements printed with inline styles.

I guess we can add those styles to the CSS file, but I'd do it in a separate PR, applying to the other instances as well 👍

ricardo commented on client/payment-request/blocks/use-initialization.js:
```diff
@@ -79,22 +82,38 @@ export const useInitialization = ( {
+	const onButtonClick = useCallback(
+		( evt, pr ) => {
+			// If login is required, display redirect confirmation dialog.
+			if ( getPaymentRequestData( 'is_login_required' ) ) {
```
Yes. [As explained above](https://github.com/Automattic/woocommerce-payments/pull/1811#discussion_r650343070). 👆

ricardo commented on client/payment-request/index.js:
```diff
@@ -186,8 +192,15 @@ jQuery( ( $ ) => {
 					return;
```
We're doing the same as in the [Stripe documentation](https://stripe.com/docs/js/payment_request/can_make_payment). I can't recall the need for a `reject` there.

Do you have any examples where it should fail?

asumaran replied:
> Do you have any examples where it should fail?

I didn't notice `canMakePayment` was a method on the Stripe SDK. It was just a non related comment.  It felt very odd to me to make an extra check on the "success" part of the promise. Better would've been to reject the promise.

asumaran commented on includes/class-wc-payments-payment-request-button-handler.php:
```diff
@@ -658,6 +720,44 @@ public function display_payment_request_button_separator_html() {
+		$redirect_url = add_query_arg(
+			[
+				'_wpnonce'                           => wp_create_nonce( 'wcpay-set-redirect-url' ),
+				'wcpay_payment_request_redirect_url' => rawurlencode( home_url( add_query_arg( [] ) ) ), // Current URL to redirect to after login.
```
@ricardo I might be confusing things but [in Stripe I worked ](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1533) on something related to this. Not entirely sure it applies to WCPay as well.

ricardo replied:
> will this work with existing parameters in the URL?

Yep. `add_query_arg` will add these args on top of the existing ones. If you open `/product/hoodie/?attribute_color=red` for example, it will redirect back to the same URL.

ricardo commented on client/payment-request/index.js:
```diff
@@ -501,22 +520,24 @@ jQuery( ( $ ) => {
 
 		attachCartPageEventListeners: ( prButton, paymentRequest ) => {
-			if (
-				( ! wcpayPaymentRequestParams.button.is_custom ||
```
In this file it's still used in `createPaymentRequestButton`. It's also used in the blocks dependencies.

ricardo commented on client/payment-request/index.js:
```diff
@@ -501,22 +520,24 @@ jQuery( ( $ ) => {
+					wcpayPaymentRequest.isCustomPaymentRequestButton(
 						prButton
-					) ) &&
-				( ! wcpayPaymentRequestParams.button.is_branded ||
```
[Same](https://github.com/Automattic/woocommerce-payments/pull/1811/files#r651134049) as for `is_custom`.

ricardo commented:
@asumaran After another design discussion I removed the Thickbox implementation and replaced it with the browser native `confirm` dialog, which reduced a lot of the boilerplate and simplified the PR.

Changes were done in 3a5cc9c73eb53c827370dec98b780dd356541e9c.

asumaran reviewed (approved):
LGTM

ricardo commented:
@asumaran I'm a bit busy this week with tasks from a different project, but it's in my next step to port this over to Stripe ([Stripe#1537](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1537)).
